src/runtime/graph/graph_runtime.cc

```diff
@@ -155,6 +155,30 @@ void GraphRuntime::SetInputZeroCopy(int index, DLTensor* data_ref) {
     t->data = data_ref->data;
   }
 }
+/*!
+ * \brief set index-th output to the graph.
+ * \param index The output index.
+ * \param data_ref The output data that is written.
+ */
+void GraphRuntime::SetOutput(int index, DLTensor* data_out) {
+  CHECK_LT(static_cast<size_t>(index), outputs_.size());
+  uint32_t eid = this->entry_id(outputs_[index]);
+  const DLTensor* old_t = data_entry_[eid].operator->();
+
+  // check the consistency of input
+  CHECK_EQ(data_alignment_[eid], details::GetDataAlignment(*data_out));
+  CHECK_EQ(reinterpret_cast<size_t>(data_out->data) % kAllocAlignment, 0);
+  CHECK_EQ(old_t->ndim, static_cast<size_t>(data_out->ndim));
+  CHECK_EQ(old_t->ctx.device_type, data_out->ctx.device_type);
+  CHECK_EQ(old_t->ctx.device_id, data_out->ctx.device_id);
+  for (auto i = 0; i < data_out->ndim; ++i) {
+    CHECK_EQ(old_t->shape[i], data_out->shape[i]);
+  }
+
+  const_cast<DLTensor*>(old_t)->data = data_out->data;
```

Comment:
When we set the pointer here, how do we prevent TVM from freeing it later since its owned by the application (and not TVM)?

And how do we free the memory already allocated by TVM for this pointer?

Reply:
This problem was also mentioned in the original PR for TVM upstream. Confusion over the owner of the tensor is introduced by this assignment. I have not worked out a solution to solve this problem well. Probably we need to resume the pointer to its original value after run. Would like to know if you have any suggestions.

Reply:
I also found that the same is happening in the SetInputZeroCopy function, where the pointer is set but theres no way to tell TVM not to free it:

https://github.com/neo-ai/tvm/blob/6363edca0e294031a670c6ec48d1e77718ee4fb3/src/runtime/graph/graph_runtime.cc#L164-L167

Reply:
I think we may need to try and use DLManagedTensor instead of DLTensor so that we can have the deleter owned by the application instead of TVM.

https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h#L157-L170